client/src/connection_cache.rs

```diff
@@ -0,0 +1,34 @@
+use {
+    crate::{tpu_connection::TpuConnection, udp_client::UdpTpuConnection},
+    lazy_static::lazy_static,
+    std::{
+        collections::HashMap,
+        net::{SocketAddr, UdpSocket},
+        sync::{Arc, Mutex},
+    },
+};
+
+lazy_static! {
+    // TODO: all implementations of TpuConnection should be Sync + Send but make sure...
+    // TODO: figure out a way to get rid of this mutex if possible
+    // try_insert seems promising but it's still nightly-only...
+    static ref CONNECTION_MAP: Mutex<HashMap<SocketAddr, Arc<dyn TpuConnection + 'static + Sync + Send>>> = Mutex::new(HashMap::new());
```

Comment:
prefer not to use lazy_static here, but maybe we can change it later if it's significantly easier.

Reply:
Agreed, the use of lazy_static here isn't ideal... will have a followup PR to improve this and other aspects.